File: backend/app/core/validation.py
"""Standard validation utilities for the Phygital backend.

Enforces Sri Lanka National Identity Card (NIC) formats, mobile phone conventions,
transaction financial limits, and secure file upload MIME validation.
"""

from __future__ import annotations

import re
from typing import Literal

# ── Sri Lankan Phone ─────────────────────────────────────────────────────────

_PHONE_RE = re.compile(r"^(?:\+?94|0)?(7\d{8})$")


def validate_sri_lankan_phone(raw: str) -> str:
    """Validate and normalize a Sri Lankan mobile number.

    Accepts formats:
        - 07XXXXXXXX (10 digits)
        - +947XXXXXXXX / 947XXXXXXXX (international)

    Returns:
        Canonical 10-digit format: ``07XXXXXXXX``.

    Raises:
        ValueError: If the number is not a valid Sri Lankan mobile number.
    """
    cleaned = raw.strip().replace(" ", "").replace("-", "")
    match = _PHONE_RE.match(cleaned)
    if not match:
        raise ValueError(
            "Phone number must be a 10-digit Sri Lankan mobile number starting with 07 (e.g. 0771234567)."
        )
    return f"0{match.group(1)}"


# ── Sri Lankan NIC ───────────────────────────────────────────────────────────

_OLD_NIC_RE = re.compile(r"^\d{9}[VX]$")
_NEW_NIC_RE = re.compile(r"^\d{12}$")


def validate_sri_lankan_nic(raw: str) -> tuple[str, Literal["male", "female"]]:
    """Validate Sri Lankan National Identity Card (NIC) and detect gender.

    Old format: 9 digits followed by 'V' or 'X' (e.g. ``896543456V``).
    New format: 12 digits (e.g. ``198965434567``).

    Returns:
        A tuple of ``(canonical_nic, gender)`` where gender is 'male' or 'female'.

    Raises:
        ValueError: If format is invalid or birth day-of-year code is out of range.
    """
    nic = raw.strip().upper()

    if len(nic) == 10 and _OLD_NIC_RE.match(nic):
        day_code = int(nic[2:5])
    elif len(nic) == 12 and _NEW_NIC_RE.match(nic):
        day_code = int(nic[4:7])
    else:
        raise ValueError(
            "NIC must be either 9 digits followed by V/X (old format) or 12 digits (new format)."
        )

    if 501 <= day_code <= 866:
        gender: Literal["male", "female"] = "female"
    elif 1 <= day_code <= 366:
        gender = "male"
    else:
        raise ValueError(f"NIC contains an invalid day-of-year code: {day_code}.")

    return nic, gender


# ── Financial Amounts ────────────────────────────────────────────────────────

MAX_TRANSACTION_AMOUNT_LKR = 100_000_000.0  # 100 Million LKR


def validate_transaction_amount(amount: float) -> float:
    """Validate that a transaction amount is positive and within acceptable bounds.

    Returns:
        Amount rounded to 2 decimal places.

    Raises:
        ValueError: If amount <= 0 or exceeds maximum limit.
    """
    if amount <= 0:
        raise ValueError("Transaction amount must be strictly greater than zero.")
    if amount > MAX_TRANSACTION_AMOUNT_LKR:
        raise ValueError(
            f"Transaction amount cannot exceed LKR {MAX_TRANSACTION_AMOUNT_LKR:,.0f}."
        )
    return round(amount, 2)


# ── File Uploads ─────────────────────────────────────────────────────────────

ALLOWED_IMAGE_EXTENSIONS = {".jpg", ".jpeg", ".png", ".webp"}
ALLOWED_AUDIO_EXTENSIONS = {".webm", ".wav", ".mp3", ".m4a", ".ogg", ".mp4"}
ALLOWED_EXTENSIONS = ALLOWED_IMAGE_EXTENSIONS | ALLOWED_AUDIO_EXTENSIONS

ALLOWED_MIME_TYPES = {
    # Images
    "image/jpeg",
    "image/png",
    "image/webp",
    # Audio & video containers
    "audio/webm",
    "video/webm",
    "audio/wav",
    "audio/x-wav",
    "audio/wave",
    "audio/mpeg",
    "audio/mp3",
    "audio/m4a",
    "audio/x-m4a",
    "audio/ogg",
    "audio/mp4",
    "video/mp4",
    "application/octet-stream",  # mobile browsers sometimes set generic mime
}


def validate_uploaded_file(filename: str, content_type: str | None = None) -> None:
    """Ensure an uploaded file matches allowed image or audio formats.

    Raises:
        ValueError: If file extension or MIME type is rejected.
    """
    if not filename:
        raise ValueError("Uploaded file must have a valid filename.")

    dot_index = filename.rfind(".")
    if dot_index == -1:
        raise ValueError(f"File '{filename}' lacks a file extension.")

    ext = filename[dot_index:].lower()
    if ext not in ALLOWED_EXTENSIONS:
        allowed_list = ", ".join(sorted(ALLOWED_EXTENSIONS))
        raise ValueError(
            f"Unsupported file format '{ext}' for file '{filename}'. Allowed formats: {allowed_list}."
        )

    if content_type and content_type.lower() not in ALLOWED_MIME_TYPES:
        raise ValueError(
            f"File '{filename}' has an unsupported content type: '{content_type}'."
        )

File: frontend/src/utils/validation.ts
/**
 * Standard client-side validation utilities for the Phygital platform.
 * Conforms to Sri Lanka National Identity Card (NIC) standards and CBSA/NCGI banking standards.
 */

export interface ValidationResult<T = void> {
  isValid: boolean;
  error?: string;
  data?: T;
}

export interface NicValidationResult {
  isValid: boolean;
  format: "old" | "new" | null;
  gender: "male" | "female" | null;
  normalized: string;
  error?: string;
}

/**
 * Validate Sri Lankan National Identity Card (NIC).
 * - Old format: 9 digits followed by 'V' or 'X' (e.g. 896543456V).
 * - New format: 12 digits (e.g. 198965434567).
 * - Day-code: 1-366 (Male), 501-866 (Female).
 */
export function validateNIC(raw: string): NicValidationResult {
  const nic = raw.trim().toUpperCase();

  if (!nic) {
    return { isValid: false, format: null, gender: null, normalized: "", error: "NIC is required." };
  }

  let format: "old" | "new" | null = null;
  let dayCode = 0;

  if (nic.length === 10 && /^\d{9}[VX]$/.test(nic)) {
    format = "old";
    dayCode = parseInt(nic.substring(2, 5), 10);
  } else if (nic.length === 12 && /^\d{12}$/.test(nic)) {
    format = "new";
    dayCode = parseInt(nic.substring(4, 7), 10);
  } else {
    return {
      isValid: false,
      format: null,
      gender: null,
      normalized: nic,
      error: "NIC must be 9 digits followed by V/X (old format) or 12 digits (new format).",
    };
  }

  let gender: "male" | "female" | null = null;
  if (dayCode >= 501 && dayCode <= 866) {
    gender = "female";
  } else if (dayCode >= 1 && dayCode <= 366) {
    gender = "male";
  } else {
    return {
      isValid: false,
      format,
      gender: null,
      normalized: nic,
      error: "Invalid birth day-code in NIC.",
    };
  }

  return {
    isValid: true,
    format,
    gender,
    normalized: nic,
  };
}

/**
 * Validate and normalize a Sri Lankan mobile number.
 * Accepts formats: 07XXXXXXXX, +947XXXXXXXX, 947XXXXXXXX.
 * Normalizes to standard 10-digit format: 07XXXXXXXX.
 */
export function validateSriLankanPhone(raw: string): ValidationResult<string> {
  const cleaned = raw.trim().replace(/[\s-]/g, "");

  if (!cleaned) {
    return { isValid: false, error: "Phone number is required." };
  }

  // Regex matching 07XXXXXXXX, +947XXXXXXXX, or 947XXXXXXXX
  const match = cleaned.match(/^(?:\+?94|0)?(7\d{8})$/);
  if (!match) {
    return {
      isValid: false,
      error: "Phone number must be a 10-digit Sri Lankan mobile number starting with 07 (e.g. 0771234567).",
    };
  }

  const normalized = `0${match[1]}`;
  return {
    isValid: true,
    data: normalized,
  };
}

/**
 * Validate borrower full name.
 * Length 2-100 chars, letters, spaces, dots, hyphens, and apostrophes.
 */
export function validateFullName(raw: string): ValidationResult<string> {
  const trimmed = raw.trim();

  if (!trimmed) {
    return { isValid: false, error: "Full name is required." };
  }

  if (trimmed.length < 2) {
    return { isValid: false, error: "Full name must be at least 2 characters." };
  }

  if (trimmed.length > 100) {
    return { isValid: false, error: "Full name cannot exceed 100 characters." };
  }

  // Unicode letters, dots, hyphens, spaces, apostrophes
  const validChars = /^[\p{L}\s.'-]+$/u;
  if (!validChars.test(trimmed)) {
    return { isValid: false, error: "Full name contains invalid characters." };
  }

  return { isValid: true, data: trimmed };
}

/**
 * Validate password.
 * Minimum 6 characters (up to 128 chars).
 */
export function validatePassword(password: string): ValidationResult<string> {
  if (!password) {
    return { isValid: false, error: "Password is required." };
  }

  if (password.length < 6) {
    return { isValid: false, error: "Password must be at least 6 characters." };
  }

  if (password.length > 128) {
    return { isValid: false, error: "Password cannot exceed 128 characters." };
  }

  return { isValid: true, data: password };
}

/**
 * Validate transaction amount in LKR.
 * Must be a positive finite number, max 100,000,000 LKR, max 2 decimal places.
 */
export function validateAmount(raw: string | number): ValidationResult<number> {
  const str = String(raw).trim();

  if (!str) {
    return { isValid: false, error: "Amount is required." };
  }

  const num = Number(str);
  if (!Number.isFinite(num) || isNaN(num)) {
    return { isValid: false, error: "Please enter a valid numeric amount." };
  }

  if (num <= 0) {
    return { isValid: false, error: "Amount must be greater than zero." };
  }

  if (num > 100_000_000) {
    return { isValid: false, error: "Amount exceeds maximum limit of LKR 100,000,000." };
  }

  // Check decimal places
  const parts = str.split(".");
  if (parts.length > 1 && parts[1].length > 2) {
    return { isValid: false, error: "Amount cannot have more than 2 decimal places." };
  }

  return { isValid: true, data: Math.round(num * 100) / 100 };
}

/**
 * Validate transaction description.
 * Length between 1 and 500 characters.
 */
export function validateDescription(raw: string): ValidationResult<string> {
  const trimmed = raw.trim();

  if (!trimmed) {
    return { isValid: false, error: "Description is required." };
  }

  if (trimmed.length > 500) {
    return { isValid: false, error: "Description cannot exceed 500 characters." };
  }

  return { isValid: true, data: trimmed };
}

/**
 * Validate 6-digit OTP code.
 */
export function validateOtpCode(raw: string): ValidationResult<string> {
  const trimmed = raw.trim();

  if (!trimmed) {
    return { isValid: false, error: "OTP code is required." };
  }

  if (!/^\d{6}$/.test(trimmed)) {
    return { isValid: false, error: "OTP code must be exactly 6 numeric digits." };
  }

  return { isValid: true, data: trimmed };
}

/**
 * Validate Bank Verification Code.
 * Standard format: PHYG-XXXX-XXXX (8 alphanumeric characters).
 */
export function validateVerificationCode(raw: string): ValidationResult<string> {
  // Strip 'phyg-', hyphens, spaces
  const cleaned = raw.replace(/^phyg-/i, "").replace(/[^a-zA-Z0-9]/g, "").toUpperCase();

  if (cleaned.length !== 8) {
    return {
      isValid: false,
      error: "Verification code must be 8 alphanumeric characters (e.g. PHYG-XXXX-XXXX).",
    };
  }

  const formatted = `PHYG-${cleaned.slice(0, 4)}-${cleaned.slice(4, 8)}`;
  return { isValid: true, data: formatted };
}
